lib/plug/conn.ex

```diff
@@ -1388,6 +1366,46 @@ defmodule Plug.Conn do
   defp adapter_inform(%Conn{adapter: {adapter, payload}}, status, headers),
     do: adapter.inform(payload, status, headers)
 
+  @doc """
+  Request a protocol upgrade from the underlying HTTP adapter.
+
+  The precise semantics of an upgrade are deliberately left unspecified here in order to
+  support arbitrary upgrades, even to protocols which may not exist today. The primary intent of
+  this function is solely to allow an application to issue an upgrade request, not to manage how
+  a given protocol upgrade takes place or what APIs the application must support in order to serve
+  this updated protocol. For details in this regard, consult the documentation of the underlying
+  adapter (such a Plug.Cowboy or Bandit).
+
+  Takes an argument describing the requested upgrade (for example, `:websocket`), and an argument
+  which contains arbitrary data which the underlying adapter is expected to interpret in the
+  context of the requested upgrade.
+
+  If the upgrade is accepted by the adapter, the returned `Plug.Conn` will have a `state` of
+  `:upgraded`. This state is considered equivalently to a 'sent' state, and is subject to the same
+  limitation on subsequent mutating operations. Note that there is no guarantee or expectation
+  that the actual upgrade process is undertaken within this function; it is entirely possible that
+  the server will only do the actual upgrade later in the connection lifecycle.
+
+  If the adapter does not support the requested upgrade then this is a noop and the returned 
+  `Plug.Conn` will be unchanged. The application can detect this and operate on the conn as it
+  normally would in order to indicate an upgrade failure to the client.
+  """
+  @spec upgrade_adapter(t, atom, term) :: t
+  def upgrade_adapter(%Conn{adapter: {adapter, payload}, state: state} = conn, protocol, opts)
```

Comment:
The _adapter suffix is intentional, as it signals to the caller is going somewhat 'off road' here by calling down into the underlying adapter in an opaque way

Reply:
I want the word `adapter` in the function because the semantics of the upgrade (the third argument) are entirely up to the adapter.

In fact, I have a nitpick that we should call it `arg` or perhaps `adapter_arg` instead of `opts` to avoid confusion?

Reply:
That's a good change.

Reply:
Updated.